service/debugger/debugger.go

```diff
@@ -0,0 +1,453 @@
+func (d *Debugger) Run() error {
+	// We must ensure here that we are running on the same thread during
+	// the execution of dbg. This is due to the fact that ptrace(2) expects
+	// all commands after PTRACE_ATTACH to come from the same thread.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Create the process by either attaching or launching.
+	var process *proctl.DebuggedProcess
+	if d.config.AttachPid > 0 {
+		glog.Infof("attaching to pid %d", d.config.AttachPid)
+		p, err := proctl.Attach(d.config.AttachPid)
+		if err != nil {
+			return fmt.Errorf("couldn't attach to pid %d: %s", d.config.AttachPid, err)
+		}
+		process = p
+	} else {
+		glog.Infof("launching process with args: %v\n", d.config.ProcessArgs)
+		p, err := proctl.Launch(d.config.ProcessArgs)
+		if err != nil {
+			return fmt.Errorf("couldn't launch process: %s\n", err)
+		}
+		process = p
+	}
+
+	// Handle access to the process from the current thread.
+	for {
+		select {
+		case f := <-d.processFuncs:
+			// Execute the function
+			f(process)
+		case s := <-d.stop:
+			// Handle shutdown
+			glog.Info("debugger is stopping")
+
+			// Clear breakpoints
+			bps := []*proctl.BreakPoint{}
+			for _, bp := range process.BreakPoints {
+				if bp != nil {
+					bps = append(bps, bp)
+				}
+			}
+			for _, bp := range process.HWBreakPoints {
+				if bp != nil {
+					bps = append(bps, bp)
+				}
+			}
+			for _, bp := range bps {
+				if bp == nil {
+					continue
+				}
+				_, err := process.Clear(bp.Addr)
+				if err != nil {
+					glog.Errorf("warning: couldn't clear breakpoint @ %s", bp.Addr)
+				} else {
+					glog.Infof("cleared breakpoint @ %v", bp.Addr)
+				}
+			}
+
+			// Detach
+			if err := sys.PtraceDetach(process.Pid); err == nil {
+				glog.Info("detached from process")
+			} else {
+				glog.Errorf("couldn't detach from process: %s\n", err)
+			}
+
+			// Kill the process if requested
+			if s.KillProcess {
+				if err := process.Process.Kill(); err == nil {
+					glog.Info("killed process")
+				} else {
+					glog.Errorf("couldn't kill process: %s\n", err)
+				}
+			}
+
+			return nil
+		}
+	}
+}
+
+// Detach stops the debugger.
+func (d *Debugger) Detach(kill bool) error {
+	d.stop <- stopSignal{KillProcess: kill}
+	return nil
+}
+
+func (d *Debugger) State() (*api.DebuggerState, error) {
+	var state *api.DebuggerState
+
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		var thread *api.Thread
+		th := p.CurrentThread
+		if th != nil {
+			thread = convertThread(th)
+		}
+
+		state = &api.DebuggerState{
+			// TODO: Running
+			// TODO: BreakPoint
+			Thread: thread,
+		}
+		return nil
+	})
+
+	return state, err
+}
+
+func (d *Debugger) CreateBreakPoint(requestedBp *api.BreakPoint) (*api.BreakPoint, error) {
+	var createdBp *api.BreakPoint
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		loc := fmt.Sprintf("%s:%d", requestedBp.File, requestedBp.Line)
+		bp, breakError := p.BreakByLocation(loc)
```

Comment:
Is the assumption here that the client will be responsible for figuring out file:line even if a function name or address is specified?